src/sci/aci_sci.cc

```diff
@@ -0,0 +1,1877 @@
+                            P_space_.size());
+    } else if (!quiet_mode_) {
+        outfile->Printf("\n  Not checking for spin-completeness.");
+    }
+    // Diagonalize H in the P space
+    if (ex_alg_ == "ROOT_ORTHOGONALIZE" and root_ > 0 and cycle_ >= pre_iter_) {
+        sparse_solver_.set_root_project(true);
+        add_bad_roots(P_space_);
+        sparse_solver_.add_bad_states(bad_roots_);
+    }
+
+    if (sparse_solver_.sigma_method_ == "HZ") {
+        op_.clear_op_lists();
+        op_.clear_tp_lists();
+        op_.build_strings(P_space_);
+        op_.op_lists(P_space_);
+        op_.tp_lists(P_space_);
+    } else if (diag_method_ != Dynamic) {
+        op_.clear_op_s_lists();
+        op_.clear_tp_s_lists();
+        op_.build_strings(P_space_);
+        op_.op_s_lists(P_space_);
+        op_.tp_s_lists(P_space_);
+    }
+
+    sparse_solver_.manual_guess(false);
+    local_timer diag;
+    sparse_solver_.diagonalize_hamiltonian_map(P_space_, op_, P_evals_, P_evecs_, num_ref_roots_,
+                                               multiplicity_, diag_method_);
+    if (!quiet_mode_)
+        outfile->Printf("\n  Time spent diagonalizing H:   %1.6f s", diag.get());
+
+    // Save ground state energy
+    P_energies_.push_back(P_evals_->get(0));
+
+    if ((cycle_ > 1) and options_->get_bool("ACI_APPROXIMATE_RDM")) {
+        double diff = std::abs(P_energies_[cycle_] - P_energies_[cycle_ - 1]);
+        if (diff <= 1e-5) {
+            approx_rdm_ = true;
+        }
+    }
+
+    // Update the reference root if root following
+    if (follow_ and num_ref_roots_ > 1 and (cycle_ >= pre_iter_) and cycle_ > 0) {
+        ref_root_ = root_follow(P_ref_, P_ref_evecs_, P_space_, P_evecs_, num_ref_roots_);
+    }
+
+    // Print the energy
+    if (!quiet_mode_) {
+        outfile->Printf("\n");
+        for (int i = 0; i < num_ref_roots_; ++i) {
+            double abs_energy =
+                P_evals_->get(i) + nuclear_repulsion_energy_ + as_ints_->scalar_energy();
+            double exc_energy = pc_hartree2ev * (P_evals_->get(i) - P_evals_->get(0));
+            outfile->Printf("\n    P-space  CI Energy Root %3d        = "
+                            "%.12f Eh = %8.4f eV",
+                            i, abs_energy, exc_energy);
+        }
+        outfile->Printf("\n");
+    }
+
+    if (!quiet_mode_ and options_->get_bool("ACI_PRINT_REFS"))
+        print_wfn(P_space_, op_, P_evecs_, num_ref_roots_);
+}
+
+void AdaptiveCI_SCI::find_q_space() {
+    // Step 2. Find determinants in the Q space
+    local_timer build_space;
+    if (options_->get_bool("ACI_BATCHED_SCREENING")) {
+        find_q_space_batched(P_space_, PQ_space_, P_evals_, P_evecs_);
+    } else if (streamline_qspace_) {
+        default_find_q_space(P_space_, PQ_space_, P_evals_, P_evecs_);
+    } else {
+        find_q_space(P_space_, PQ_space_, num_ref_roots_, P_evals_, P_evecs_);
```

Comment:
This `find_q_space` should maybe have a different name to avoid confusion. Something like `find_q_space_multiroot`